### PR TITLE
LR: event listener resubscribe onError

### PR DESCRIPTION
**Pull request by shama358**
## Overview
This change is to automatically resubscribe to the event listener when the streaming task encounters an error.

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java [hunk 1]
```diff
@@ -323,6 +293,169 @@ private void triggerSnapshotAndTestRemainingEntries() throws Exception{
...
+            try(TxnContext txn = corfuStoreActive.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                assertThat(txn.getTable(REPLICATION_EVENT_TABLE_NAME).count()).isZero();
+            }
+
+            //reset the listener
+            standbyListener.reset(new CountDownLatch(totalStandbyStatusUpdates));
+
+            // Add an invalid entry to the table so the event listener's onError() gets called in a loop, giving the test to CP/trim
+            try(TxnContext txn = corfuStoreActive.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                txn.putRecord(replicationEventTable, key, LogReplicationMetadata.ReplicationEvent.newBuilder().build(), null);
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: will there be a failure in the listener itself or will snapshot sync eventually fail because the value is uninitialized?

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java [hunk 2]
```diff
@@ -323,6 +293,169 @@ private void triggerSnapshotAndTestRemainingEntries() throws Exception{
...
+                txn.commit();
+            }
+
+            // delete the invalid entry, so the full sync upon resubscription doesn't see the invalid record
+            try(TxnContext txn = corfuStoreActive.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                txn.delete(replicationEventTable, key);
+                txn.commit();
+            }
+
+            // run CP twice so the listener hits a trimmed exception
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:414]: the trimmed exception will be hit only if there is new data(onNext()) or resubscription.  If resubscription has already finished by the time of trim, no exception will be hit, right?

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java [hunk 3]
```diff
@@ -323,6 +293,169 @@ private void triggerSnapshotAndTestRemainingEntries() throws Exception{
         assertThat(replicationStatusVal.getRemainingEntriesToSend()).isEqualTo(0);
     }
 
+    /**
+     * This test verifies that the eventListener is resubscribed correctly when the streaming layer encounters an error.
+     * 1. Validate that the setup created is stable by validating the snapshot_sync data
+     * 2. Add a valid force snapshot_sync event in the event Table.
+     * 3. Ensure that the snapshot sync was completed for the event and the event is effectively removed once processed
+     * 4. Add an invalid record to the event table. This makes the eventListener in LR stall by repeatedly trying to
```
Comment (unlaphoi) [inline test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:301]: Can we also test the case where StreamingException is thrown?

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java (hunk 1 above)
Comment (shama358) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: the failure is in listener thread, while it tries to enqueue the event in the discoveryServices's queue.
Basically on `OnNext()` -> `processEvent()`

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java (hunk 2 above)
Comment (shama358) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:414]: Yes...so the invalid Entry makes sure that the listener keeps trying to read that entry, fails, resubscribes back to the same address, and again tries to read that invalid entry.

So the listerner is stuck in that loop...and then we delete that invalid entry and run CP twice so that the invalid entry is effectively deleted. Since listener is still stuck in the above loop, it hits a trimmed exception

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java (hunk 1 above)
Comment (pankti-m) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: unable to follow how it will fail.. because everything in the event protobuf will have default values, including the event type and it will get added to the queue.  Am I missing something?

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java (hunk 2 above)
Comment (pankti-m) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:414]: yes.. its related to my question in the first comment.  I am unable to follow where the empty event causes a failure in the listener

File: test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java (hunk 1 above)
Comment (shama358) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: yes, but it fails while trying to convert to UUID. This is the error with the stack trace:


```
2024-04-08T17:45:23.312Z | ERROR | treamPollingScheduler-worker-0 | r.c.StreamListenerResumePolicy | Exception caught during streaming processing. Re-subscribe this listener to latest timestamp. Error=
java.lang.IllegalArgumentException: Invalid UUID string:
        at java.base/java.util.UUID.fromString(UUID.java:215)
        at org.corfudb.infrastructure.logreplication.infrastructure.DiscoveryServiceEvent.<init>(DiscoveryServiceEvent.java:35)
        at org.corfudb.infrastructure.logreplication.infrastructure.LogReplicationEventListener.processEvent(LogReplicationEventListener.java:96)
        at org.corfudb.infrastructure.logreplication.infrastructure.LogReplicationEventListener.onNext(LogReplicationEventListener.java:85)
        at org.corfudb.runtime.collections.StreamListenerResumePolicy.onNextEntry(StreamListenerResumePolicy.java:75)
        at org.corfudb.runtime.collections.StreamListenerResumeOrFullSync.onNextEntry(StreamListenerResumeOrFullSync.java:20)
        at org.corfudb.runtime.collections.streaming.StreamingTask.lambda$produce$4(StreamingTask.java:194)
        at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:79)
        at org.corfudb.common.metrics.micrometer.MicroMeterUtils.time(MicroMeterUtils.java:115)
        at org.corfudb.runtime.collections.streaming.StreamingTask.lambda$produce$5(StreamingTask.java:194)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at org.corfudb.runtime.collections.streaming.StreamingTask.produce(StreamingTask.java:194)
        at org.corfudb.runtime.collections.streaming.StreamingTask.run(StreamingTask.java:223)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

Comment (shama358) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: So the listener is gets resubscribed and then again hits this...keeping it in this loop until we run CP/trim twice

Comment (pankti-m) [reply test/src/test/java/org/corfudb/integration/LogReplicationAbstractIT.java:404]: got it.. can you add this in the comments or test description (something like 'event id is empty so construction of DiscoveryServiceEvent throws an exception')?  This is for the reader to be sure that the resubscription will indeed happen.

**Review by pankti-m — APPROVED** [review]

**Review by unlaphoi — APPROVED** [review]

**Review by shreayp — APPROVED** [review]
